engine/runtime/src/main/java/org/enso/interpreter/runtime/data/hash/EnsoHashMapBuilder.java

```diff
@@ -1,193 +1,264 @@
+        if (invalidatedEntry != byHash[at]) {
+          byHash[at] = invalidatedEntry;
+          actualSize--;
+          return true;
+        }
+      }
+      if (++at == byHash.length) {
+        at = 0;
       }
-      size--;
-      return true;
     }
+    throw CompilerDirectives.shouldNotReachHere("byHash array is full!");
   }
 
-  @TruffleBoundary(allowInlining = true)
-  public boolean containsKey(Object key) {
-    return storage.containsKey(key);
+  /** Builds a new builder with given array size and puts into it all entries
+   * that are valid {@code atGeneration}.
+   */
+  private EnsoHashMapBuilder rehash(int size, int atGeneration, HashCodeNode hashCodeNode, EqualsNode equalsNode) {
+    var newBuilder = new EnsoHashMapBuilder(size);
+    for (var i = 0; i < byHash.length; i++) {
+      var entry = byHash[i];
+      if (entry != null && entry.isVisible(atGeneration)) {
+        newBuilder.put(entry.key(), entry.value(), hashCodeNode, equalsNode);
+      }
+    }
+    return newBuilder;
   }
 
+
   /**
    * Creates a snapshot with the current size. The created snapshot contains all the entries that
    * are in the storage as of this moment, i.e., all the entries with their indexes lesser than
-   * {@code size}.
+   * {@code generation}.
    *
-   * <p>Should be called at most once for a particular {@code size}.
+   * <p>Should be called where most once for a particular {@code generation}.
    *
    * @return A new hash map snapshot.
    */
   public EnsoHashMap build() {
-    return EnsoHashMap.createWithBuilder(this, size);
+    return EnsoHashMap.createWithBuilder(this);
   }
 
   @Override
   public String toString() {
-    return "EnsoHashMapBuilder{size = " + size + ", storage = " + storage + "}";
+    return "EnsoHashMapBuilder{size = " + generation + ", storage = " + Arrays.toString(byHash) + "}";
   }
 
-  record StorageEntry(
-      Object key,
-      Object value,
-      /**
-       * A sequential index of the entry within this map. {@link EnsoHashMap} uses it for checking
-       * whether a certain key belongs in that map.
-       */
-      int index) {}
+  private static boolean compare(EqualsNode equalsNode, Object a, Object b) {
+    if (a instanceof Double aDbl && b instanceof Double bDbl && aDbl.isNaN() && bDbl.isNaN()) {
+      return true;
+    } else {
+      return equalsNode.execute(a, b);
+    }
+  }
 
-  /**
-   * Custom {@link Equivalence} used for the {@link EconomicMap} that delegates {@code equals} to
-   * {@link EqualsNode} and {@code hash_code} to {@link HashCodeNode}.
-   */
-  private static final class StorageStrategy extends Equivalence {
-    private final EqualsNode equalsNode;
-    private final HashCodeNode hashCodeNode;
+  record StorageEntry(
+    Object key,
+    Object value,
+    /**
+    * A sequential index of the entry within this map. {@link EnsoHashMap} uses it for checking
+    * whether a certain key belongs in that map.
+    */
+    int added,
```

Comment:
Seems like a leftover javadoc? Is it up to date?

Reply:
Changed in 2c3100f2b7